concent_api/core/payments/sci_callback.py

```diff
@@ -0,0 +1,260 @@
+from core.exceptions import SCICallbackTransactionSignatureError
+
+from middleman_protocol.concent_golem_messages.message import SignedTransaction
+from middleman_protocol.concent_golem_messages.message import TransactionRejected
+from middleman_protocol.concent_golem_messages.message import TransactionSigningRequest
+from middleman_protocol.exceptions import MiddlemanProtocolError
+from middleman_protocol.message import ErrorFrame
+from middleman_protocol.message import GolemMessageFrame
+from middleman_protocol.stream import send_over_stream
+from middleman_protocol.stream import unescape_stream
+
+
+class RequestIDGenerator:
+
+    _request_id = 0
+
+    @classmethod
+    def generate_request_id(cls) -> int:
+        cls._request_id += 1
+        return cls._request_id
+
+
+def sci_callback(transaction: Transaction) -> Transaction:
+    """
+    This is callback that submits the transaction to the Signing Service
+    and does not rely on having access to the private key.
+
+    Returns signed Ethereum Transaction.
+
+    The flow is:
+    1. Create a TransactionSigningRequest.
+    2. Establish a TCP connection with Middleman.
+    3. Send the TransactionSigningRequest.
+    4. Read the response and close the connection.
+    5a. If the response is a valid SignedTransaction, copy the signature to the transaction object passed to the
+        callback by SCI and return.
+    5b. Otherwise handle errors by raising an exception and interrupting the callback.
+
+    """
+    assert isinstance(transaction, Transaction)
+
+    # Create a TransactionSigningRequest.
+    transaction_signing_request = create_transaction_signing_request(transaction)
+
+    # Generate request ID.
+    request_id = RequestIDGenerator.generate_request_id()
+
+    # Create a MiddleMan Protocol Golem Message Frame.
+    middleman_message = GolemMessageFrame(
+        payload=transaction_signing_request,
+        request_id=request_id,
+    )
+
+    # Send Frame to MiddleMan through MiddleMan Protocol and receive response.
+    raw_response = send_request_to_middleman(middleman_message)
+
+    # Deserialize received Frame and its payload and handle related errors.
+    signed_transaction = deserialize_response_and_handle_errors(raw_response, request_id)
+
+    assert isinstance(signed_transaction, SignedTransaction)
+
+    # Verify received SignedTransaction signature and handle related errors.
+    verify_data_and_signature(signed_transaction, transaction)
+
+    # If the response is a valid SignedTransaction, copy Ethereum Transaction object into new one and
+    # add SignedTransaction data.
+    return copy_transaction_and_signature(signed_transaction, transaction)
+
+
+def create_transaction_signing_request(transaction: Transaction) -> TransactionSigningRequest:
+    """ Create TransactionSigningRequest message from given Ethereum Transaction object. """
+
+    assert isinstance(transaction, Transaction)
+
+    transaction_signing_request = TransactionSigningRequest(
```

Comment:
can we please change `from` to `from_`, so it could be used in constructor (via named params)? Right know we are having an `TransactionSigningRequest`object constructed with two steps: call to "constructor" and call to `setattr`.

Reply:
How about changing the name of it to `source` or `sender`? @cameel what do you think?

Reply:
Sure. I would personally go with `from_address`.

Please drop me a comment about it in #632 so that I remember about it when I'll be updating the docs.

Reply:
:+1: done